apps/language_server/lib/language_server/providers/workspace_symbols.ex

```diff
@@ -0,0 +1,468 @@
+defmodule ElixirLS.LanguageServer.Providers.WorkspaceSymbols do
+  @moduledoc """
+  Workspace Symbols provider. Generates and returns `SymbolInfo[]`.
+
+  https://microsoft.github.io//language-server-protocol/specifications/specification-3-14/#workspace_symbol
+  """
+  use GenServer
+
+  alias ElixirLS.LanguageServer.ErlangSourceFile
+  alias ElixirLS.LanguageServer.SourceFile
+  alias ElixirLS.LanguageServer.Providers.SymbolUtils
+  alias ElixirLS.LanguageServer.JsonRpc
+
+  @type position_t :: %{
+          line: non_neg_integer,
+          character: non_neg_integer
+        }
+
+  @type range_t :: %{
+          start: position_t,
+          end: position_t
+        }
+
+  @type location_t :: %{
+          uri: String.t(),
+          range: range_t
+        }
+
+  @type symbol_information_t :: %{
+          kind: integer,
+          name: String.t(),
+          location: location_t
+        }
+
+  @typep key_t :: :modules | :functions | :types | :callbacks
+  @typep symbol_t :: module | {module, atom, non_neg_integer}
+  @typep state_t :: %{
+           required(key_t) => [symbol_information_t],
+           modified_uris: [String.t()]
+         }
+
+  @symbol_codes for {key, kind} <- [
+                      modules: :module,
+                      functions: :function,
+                      types: :class,
+                      callbacks: :event
+                    ],
+                    into: %{},
+                    do: {key, SymbolUtils.symbol_kind_to_code(kind)}
+
+  ## Client API
+
+  @spec symbols(String.t()) :: {:ok, [symbol_information_t]}
+  def symbols(query) do
+    results =
+      case query do
+        "f " <> fun_query ->
```

Comment:
I wonder if there's some way to teach the user that there's these three prefixes? Is this common in other language servers as well?

Reply:
`iex` uses `h` (modules and functions), `t` (types), `b` (callbacks)
some web browsers also use those, e.g. `y query` would search youtube, `w query` wikipedia etc

Anyway, I'm not pushing for them but they help in two cases. Without prefixes search is noticeably slower and they increase search results' relevance.

Reply:
Yeah, for now we can just document it in the README. Perhaps in the future LSP will add some additional parameters to `workspace/symbol`.

Reply:
Added readme note in https://github.com/elixir-lsp/elixir-ls/pull/110/commits/72db8359a7383c097ae2881918520266021f3fec